src/unistd.rs

```diff
@@ -1567,8 +1567,16 @@ pub fn getgrouplist(user: &CStr, group: Gid) -> Result<Vec<Gid>> {
             // BSD systems will still fill the groups buffer with as many
             // groups as possible, but Linux manpages do not mention this
             // behavior.
-            reserve_double_buffer_size(&mut groups, ngroups_max as usize)
-                .map_err(|_| Errno::EINVAL)?;
+            // Linux stores the number of groups found in ngroups.  We can
+            // use that to resize the buffer exactly.
+            cfg_if! {
+                if #[cfg(target_os = "linux")] {
```

Comment:
I'm not sure about other platforms so this is only for linux.  I also skipped checking if `ngroups` is within `ngroups_max` in this case, to keep the code simple.  The assumption is that linux is consistent between `sysconf(NGROUPS_MAX)` and its return value here.

Reply:
It seems to be true on Solaris: https://docs.oracle.com/cd/E86824_01/html/E54766/getgrouplist-3c.html
and on musl but only since 2015: https://git.musl-libc.org/cgit/musl/commit/?id=2894a44b40e460fc4112988407818439f2e9672d

It might be better to detect this instead of trying to hard-code platforms, maybe something like
- store the old ngroups
- call getgrouplist
- if ngroups != old_ngroups, use that, otherwise use 2 * old_ngroups

Reply:
Why don't I revert this portion of the change and save it for another day? I agree that a more principled approach might be better.  For now we can focus on just the buffer overrun fix.